carousel/core/calculations.py

```diff
@@ -211,127 +128,122 @@ def calc_static(self, formula_reg, data_reg, out_reg, timestep):
+                            b = outargs[b]
+                        cov[:, m, n] = avar.get(b, 0.0)
+                if nobs == 1:
+                    cov = cov.squeeze()  # squeeze out any extra dimensions
+                LOGGER.debug('covariance:\n%r', cov)
+            data = index_registry(args, 'data', data_reg, timestep)
+            outputs = index_registry(args, 'outputs', out_reg, timestep)
+            kwargs = dict(data, **outputs)
+            args = [kwargs.pop(a) for a in fargs if a in kwargs]
+            returns = calc['returns']  # return arguments
+            # update kwargs with covariance if it exists
+            if cov is not None:
+                kwargs['__covariance__'] = cov
+            retval = func(*args, **kwargs)  # calculate function
+            # update output registry with covariance and jacobian
+            if cov is not None:
+                # split uncertainty and jacobian from return values
+                cov, jac = retval[-2:]
+                retval = retval[:-2]
+                # scale covariance
+                scale = np.asarray(
+                    [1 / r.m if isinstance(r, UREG.Quantity) else 1 / r
+                     for r in retval]
+                )  # use magnitudes if quantities
+                cov = (np.swapaxes((cov.T * scale), 0, 1) * scale).T
+                nret = len(retval)  # number of return output
+                for m in xrange(nret):
+                    a = returns[m]  # name in output registry
+                    out_reg.variance[a] = {}
+                    out_reg.uncertainty[a] = {}
+                    out_reg.jacobian[a] = {}
+                    for n in xrange(nret):
+                        b = returns[n]
+                        out_reg.variance[a][b] = cov[:, m, n]
+                        if a == b:
+                            unc = np.sqrt(cov[:, m, n]) * 100 * UREG.percent
+                            out_reg.uncertainty[a][b] = unc
+                    for n in xrange(argn):
+                        b = vargs[n]
+                        try:
+                            b = datargs[b]
+                        except (KeyError, TypeError):
+                            b = outargs[b]
+                        out_reg.jacobian[a][b] = jac[:, m, n]
+                    LOGGER.debug('%s cov:\n%r', a, out_reg.variance[a])
+                    LOGGER.debug('%s jac:\n%r', a, out_reg.jacobian[a])
+                    LOGGER.debug('%s unc:\n%r', a, out_reg.uncertainty[a])
+            # if there's only one return value, squeeze out extra dimensions
+            if len(retval) == 1:
+                retval = retval[0]
+            # put return values into output registry
+            if len(returns) > 1:
+                # more than one return, zip them up
+                out_reg.update(zip(returns, retval))
+            else:
+                # only one return, get it by index at 0
+                out_reg[returns[0]] = retval
```

Comment:
Sometimes I can't use GitHub to review these, so I do use meld a lot. Have you tried it?